### PR TITLE
introduce cpu & gpu time measurement for stages

### PR DESCRIPTION
**Pull request by chistopher**
working prototype
ToDo:

- [x] implement flag to enable/disable measurement
- [x] control measurement from canvas
- [x] write data to file

### CODE REVIEW

File: source/gloperate/include/gloperate/pipeline/Pipeline.h [hunk 1]
```diff
@@ -127,6 +127,15 @@ class GLOPERATE_API Pipeline : public Stage
     */
     void invalidateStageOrder();
 
+	/**
+	*  @brief
+	*    Set measurement flag for all stages. true(enable) false(disable)
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Pipeline.h]: Move the semantics of true and false to the description of the parameter
```
*  @param[in] flag
*    ...
```

File: source/gloperate/include/gloperate/pipeline/Stage.h [hunk 2]
```diff
@@ -652,6 +652,59 @@ class GLOPERATE_API Stage : public cppexpose::Object
     template <typename T>
     void forAllOutputs(std::function<void(gloperate::Output<T> *)> callback);
 
+	/**
+	*  @brief
+	*    The time spent in onProcess in nanoseconds.
+	*
+	*  @return
+	*    unsigned long representing cpu duration of onProcess in nanoseconds
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Stage.h]: Shorten this description to *CPU duration of onProcess in nanoseconds*

File: source/gloperate/include/gloperate/pipeline/Stage.h [hunk 3]
```diff
@@ -652,6 +652,59 @@ class GLOPERATE_API Stage : public cppexpose::Object
...
+	*    Due to the async nature of gpu processing, this value is one iteration (i.e. frame) late.
+	*/
+	std::uint64_t lastGPUTime() const;
+
+	/**
+	*  @brief
+	*    Return the history of cpu time measurement of onProcess (most recent values last).
+	*
+	*  @remarks
+	*    Complexity is linear in the size of the history.
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Stage.h]: Does this complexity count for the accessor or the internal management during measurements?

File: source/gloperate/include/gloperate/pipeline/Stage.h [hunk 4]
```diff
@@ -652,6 +652,59 @@ class GLOPERATE_API Stage : public cppexpose::Object
...
+	*    Due to the async nature of gpu processing, these values are one iteration (i.e. frame) late.
+	*    This means that the value at index n corresponds to the value at index n-1 in the cpu measurement history.
+	*    Complexity is linear in the size of the history.
+	*/
+	std::vector<std::uint64_t> historyGPU() const;
+
+	/**
+	*  @brief
+	*    Set the amount of cached values for time measurement. The size
+	*    must be even! Also clears previous history.
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Stage.h]: Add parameter description.

File: source/gloperate/include/gloperate/pipeline/Stage.h [hunk 5]
```diff
@@ -783,6 +836,13 @@ class GLOPERATE_API Stage : public cppexpose::Object
     Environment * m_environment;    ///< Gloperate environment to which the stage belongs
     bool          m_alwaysProcess;  ///< Is the stage always processed?
 
+	std::vector<std::uint64_t>  m_cpuTimes;
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Stage.h]: Add comments for this and the following member variables.

File: source/gloperate/include/gloperate/pipeline/Stage.h [hunk 6]
```diff
@@ -783,6 +836,13 @@ class GLOPERATE_API Stage : public cppexpose::Object
...
+	std::vector<std::uint64_t>  m_gpuTimes;
+	unsigned int                m_measurementCycle;
+	unsigned int                m_measurementHistorySize;
+	unsigned int                m_queryID[2][2];
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Stage.h]: Please change the type of `m_queryID` to be more safe and C++-idiomatic (e.g., `std::array<unsigned int, 4>`).

File: source/gloperate/source/pipeline/Pipeline.cpp [hunk 7]
```diff
@@ -142,6 +142,14 @@ void Pipeline::invalidateStageOrder()
     m_sorted = false;
 }
 
+void Pipeline::setMeasurementFlagForAllStages(bool flag)
+{
+	for (auto stage : stages())
+	{
+		stage->setMeasurementFlag(flag);
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Pipeline.cpp]: Change all occurrences of tabs with 4 spaces.

File: source/gloperate/source/pipeline/Stage.cpp [hunk 8]
```diff
@@ -40,7 +42,12 @@ Stage::Stage(Environment * environment, const std::string & className, const std
 : cppexpose::Object((name.empty()) ? className : name)
 , m_environment(environment)
 , m_alwaysProcess(false)
+, m_measurementCycle(0)
+, m_measurementHistorySize(10) //must be even
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: Add explanation of restriction. This information is suitable for the member documentation, too.

File: source/gloperate/source/pipeline/Stage.cpp [hunk 9]
```diff
@@ -90,6 +97,13 @@ bool Stage::requires(const Stage * stage, bool recursive) const
 void Stage::initContext(AbstractGLContext * context)
 {
     debug(2, "gloperate") << this->qualifiedName() << ": initContext";
+
+	gl::glGenQueries(2, m_queryID[0]); //front buffer
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: Use symbolic constants for 0 and 1 in the different contexts.

File: source/gloperate/source/pipeline/Stage.cpp [hunk 10]
```diff
@@ -102,8 +116,40 @@ void Stage::deinitContext(AbstractGLContext * context)
...
     debug(1, "gloperate") << this->qualifiedName() << ": processing";
-    onProcess();
 
+	if (m_enableMeasurement)
+	{
+		//cpu
+		auto start = std::chrono::high_resolution_clock::now();
+		//gpu ToDo check for opengl context
+		unsigned int queryFrontBuffer = m_measurementCycle % 2;
+		unsigned int queryBackBuffer = (m_measurementCycle + 1) % 2;
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: `unsigned int queryBackBuffer = 1 - queryFrontBuffer;`

File: source/gloperate/source/pipeline/Stage.cpp [hunk 11]
```diff
@@ -530,5 +576,61 @@ void Stage::forAllOutputs(std::function<void(gloperate::AbstractSlot *)> callbac
...
+
+std::uint64_t Stage::lastGPUTime() const
+{
+	return m_gpuTimes[m_measurementCycle - 1 % m_measurementHistorySize];
+}
+
+std::vector<std::uint64_t> Stage::historyCPU() const
+{
+	std::vector<std::uint64_t> history;
+	for (auto i = m_measurementCycle; i < m_measurementHistorySize; ++i)
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: use `std::copy` interface.

File: source/gloperate/source/pipeline/Stage.cpp [hunk 12]
```diff
@@ -530,5 +576,61 @@ void Stage::forAllOutputs(std::function<void(gloperate::AbstractSlot *)> callbac
...
+}
+
+std::uint64_t Stage::lastGPUTime() const
+{
+	return m_gpuTimes[m_measurementCycle - 1 % m_measurementHistorySize];
+}
+
+std::vector<std::uint64_t> Stage::historyCPU() const
+{
+	std::vector<std::uint64_t> history;
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: Initialize history with target size (`m_measurementHistorySize`).

File: source/gloperate/source/pipeline/Stage.cpp [hunk 13]
```diff
@@ -530,5 +576,61 @@ void Stage::forAllOutputs(std::function<void(gloperate::AbstractSlot *)> callbac
...
+	m_cpuTimes.clear(); 
+	m_cpuTimes.resize(size);
+	m_gpuTimes.clear();
+	m_gpuTimes.resize(size);
+	m_measurementCycle = m_measurementCycle % 2; //reset but keep even/odd
+}
+
+void Stage::setMeasurementFlag(bool flag)
+{
+	m_enableMeasurement = flag;
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: Add comment that explains that we do not clear the history.

**Comment by chistopher** [comment]
Thanks for the review. I adapted the code accordingly.

**Comment by chistopher** [comment]
I made an alternative implementation without lambda. (see alternative_measure branch)

File: source/gloperate/source/base/Canvas.cpp [hunk 14]
```diff
@@ -359,6 +361,9 @@ void Canvas::render(globjects::Framebuffer * targetFBO)
     // Render
     m_renderStage->process();
 
+    if (m_enableMeasurement)
```
Comment (scheibel) [inline source/gloperate/source/base/Canvas.cpp]: Please add opening and closing curly braces (use a full block).

File: source/gloperate/source/pipeline/Stage.cpp [hunk 15]
```diff
@@ -530,5 +599,51 @@ void Stage::forAllOutputs(std::function<void(gloperate::AbstractSlot *)> callbac
...
+}
+
+void Stage::setMeasurementFlag(bool flag, bool recursive)
+{
+    m_currentCPUDuration = 0;
+    m_lastCPUDuration = 0;
+    m_lastGPUDuration = 0;
+    m_enableMeasurement = flag;
+
+    if (recursive && isPipeline())
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: Please convert this method into a virtual member function on `Stage` and `Pipeline` while handling the substage-iteration on within the `Pipeline`.

File: source/gloperate/source/pipeline/Stage.cpp [hunk 16]
```diff
@@ -530,5 +599,51 @@ void Stage::forAllOutputs(std::function<void(gloperate::AbstractSlot *)> callbac
...
+    {
+        for (auto stage : static_cast<Pipeline*>(this)->stages())
+        {
+            stage->setMeasurementFlag(flag, true);
+        }
+    }
+}
+
+void Stage::sendMeasurementValues(std::function<void(Stage*, uint64_t, uint64_t)> func)
```
Comment (scheibel) [inline source/gloperate/source/pipeline/Stage.cpp]: Same here.

File: source/gloperate/source/tools/MeasurementExporter.cpp [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    m_canvas->setMeasurement(false);
+    writeData();
+    m_data.clear();
+    m_stages.clear();
+    m_connection = cppexpose::ScopedConnection();
+}
+
+void MeasurementExporter::receiveValue(Stage * stage, int cycle, uint64_t cpu_time, uint64_t gpu_time)
+{
+    if (m_data[stage].size() >= 1000)
```
Comment (scheibel) [inline source/gloperate/source/tools/MeasurementExporter.cpp]: Extract into named constant.

File: source/gloperate/source/tools/MeasurementExporter.cpp [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        writeData();
+
+    m_data[stage].push_back({ cpu_time, gpu_time });
```
Comment (scheibel) [inline source/gloperate/source/tools/MeasurementExporter.cpp]: You may use `emplace_back`.

File: source/gloperate/source/tools/MeasurementExporter.cpp [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+void MeasurementExporter::writeData()
+{
+    std::ofstream file;
+    file.open("temp3782099.csv", std::ios::out | std::ios::app);
```
Comment (scheibel) [inline source/gloperate/source/tools/MeasurementExporter.cpp]: This seems to be a temporary test filename.

File: source/gloperate/source/tools/MeasurementExporter.cpp [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    //clear data
+    for (auto& each : m_data)
+    {
+        each.second.clear();
+    }
+}
+
+
+
+} //namespace gloperate
```
Comment (scheibel) [inline source/gloperate/source/tools/MeasurementExporter.cpp]: Missing newline.

**Comment by chistopher** [comment]
replaced by #410

**Pull request closed without merging**